src/ObjectUtils/CoffFile.cpp

```diff
@@ -159,96 +154,158 @@ void CoffFileImpl::AddNewDebugSymbolsFromCoffSymbolTable(
+  for (SymbolInfo& symbol_info : *symbol_infos) {
+    ORBIT_CHECK(symbol_info.size() == SymbolsFile::kUnknownSymbolSize);
+    auto unwind_range_start_to_size_it = unwind_range_start_to_size.find(symbol_info.address());
+    if (unwind_range_start_to_size_it != unwind_range_start_to_size.end()) {
+      symbol_info.set_size(unwind_range_start_to_size_it->second);
     }
-
-    new_symbol_infos.emplace_back(std::move(symbol_info));
   }
-
-  ORBIT_LOG(
-      "Added %u function symbols from COFF symbol table on top of %u from DWARF debug info for "
-      "\"%s\"",
-      new_symbol_infos.size(), symbol_infos->size(), file_path_.string());
-  symbol_infos->insert(symbol_infos->end(), std::make_move_iterator(new_symbol_infos.begin()),
-                       std::make_move_iterator(new_symbol_infos.end()));
 }
 
-void FillDebugSymbolsFromDwarf(llvm::DWARFContext* dwarf_context,
-                               std::vector<SymbolInfo>* symbol_infos) {
-  for (const auto& info_section : dwarf_context->compile_units()) {
-    for (uint32_t index = 0; index < info_section->getNumDIEs(); ++index) {
-      llvm::DWARFDie full_die = info_section->getDIEAtIndex(index);
+void CoffFileImpl::AddNewDebugSymbolsFromDwarf(llvm::DWARFContext* dwarf_context,
+                                               std::vector<SymbolInfo>* symbol_infos) {
+  // Sort so that we can use std::lower_bound below.
+  std::sort(symbol_infos->begin(), symbol_infos->end(), &SymbolsFile::SymbolInfoLessByAddress);
+
+  std::vector<SymbolInfo> new_symbol_infos;
+  for (const auto& compile_unit : dwarf_context->compile_units()) {
+    for (uint32_t die_index = 0; die_index < compile_unit->getNumDIEs(); ++die_index) {
+      llvm::DWARFDie full_die = compile_unit->getDIEAtIndex(die_index);
       // We only want symbols of functions, which are DIEs with isSubprogramDIR(),
       // and not of inlined functions, which are DIEs with isSubroutineDIE().
-      if (full_die.isSubprogramDIE()) {
-        uint64_t low_pc;
-        uint64_t high_pc;
-        uint64_t unused_section_index;
-        if (!full_die.getLowAndHighPC(low_pc, high_pc, unused_section_index)) {
+      if (!full_die.isSubprogramDIE()) {
+        continue;
+      }
+
+      uint64_t low_pc{};
+      uint64_t high_pc{};
+      uint64_t unused_section_index{};
+      // For some functions in some Wine DLLs (such as d3d9.dll, d3d11.dll, dxgi.dll) we get a zero
+      // address even if getLowAndHighPC succeeds: skip such functions.
+      if (!full_die.getLowAndHighPC(low_pc, high_pc, unused_section_index) || low_pc == 0) {
+        continue;
+      }
+
+      auto symbol_from_symbol_table_it =
+          std::lower_bound(symbol_infos->begin(), symbol_infos->end(), low_pc,
+                           [](const orbit_grpc_protos::SymbolInfo& lhs, uint64_t rhs) {
+                             return lhs.address() < rhs;
+                           });
+      if (symbol_from_symbol_table_it != symbol_infos->end() &&
```

Comment:
What about the case, where the symbol in the symbol table does not have a size,yet?

Reply:
I was more wondering, if we want to "patch" the sizes of those, as we do have the information here?

Reply:
I see, good observation. The thing is that (somewhat explained in the big comment on `LoadDebugSymbols`), from what I have observed, all functions from DWARF information are either already in the COFF symbol table *and* have a RUNTIME_FUNCTION, or are not in the COFF symbol table at all. However, it makes sense to add this case for completeness: added it with a comment.